### ISSUE TITLE
CTC backprop gradient efficiency

### ISSUE THREAD

**Issue by naxingyu** [srvk/eesen#5]
Hi Yajie,
I notice that in the ctc code, the gradient of output is calculated and the then multiplied with the nnet output to get the backprop error signal. Referring to Alex's thesis, it seems that you implemented Eq.(7.31) first, then implemented Eq.(7.32). Is it possible to implement Eq(7.34) directly such that there will be one matrix multiplication less?

**Comment by ghost** [comment]
hi Xingyu,
Do you mean Eq. 7.26 and 7.28 in the following thesis: http://www.cs.toronto.edu/~graves/phd.pdf?
My implementing CTC evaluation and softmax-layer back-propagation separately indeed results in some redundancy. For instance, Eq. 7.26 is divided by y_t^k which is later multiplied back in computation softmax BP. Is this what you refer to?

**Comment by naxingyu** [comment]
Hi Yajie,
Exactly. I don't if it works if you implement Eq. 7.29 directly? The 
devision by y_t^k imposes the using of a question operator. Maybe it's 
necessary for certain conditions?
Xingyu
